datafusion/optimizer/src/utils.rs

```diff
@@ -99,27 +99,176 @@ fn split_conjunction_impl<'a>(expr: &'a Expr, mut exprs: Vec<&'a Expr>) -> Vec<&
+///
+/// For example, given the input `[[a, b], [c], [d, e]]` returns
+/// `[a, c, d], [a, c, e], [b, c, d], [b, c, e]]`.
+fn permutations(mut exprs: VecDeque<Vec<&Expr>>) -> Vec<Vec<&Expr>> {
+    let first = if let Some(first) = exprs.pop_front() {
+        first
+    } else {
+        return vec![];
+    };
+
+    // base case:
+    if exprs.is_empty() {
+        first.into_iter().map(|e| vec![e]).collect()
+    } else {
+        first
+            .into_iter()
+            .flat_map(|expr| {
+                permutations(exprs.clone())
+                    .into_iter()
+                    .map(|expr_list| {
+                        // Create [expr, ...] for each permutation
+                        std::iter::once(expr)
+                            .chain(expr_list.into_iter())
+                            .collect::<Vec<&Expr>>()
+                    })
+                    .collect::<Vec<Vec<&Expr>>>()
+            })
+            .collect()
+    }
+}
+
+const MAX_CNF_REWRITE_CONJUNCTS: usize = 10;
+
+/// Tries to convert an expression to conjunctive normal form (CNF).
+///
+/// Does not convert the expression if the total number of conjuncts
+/// (exprs ANDed together) would exceed [`MAX_CNF_REWRITE_CONJUNCTS`].
+///
+/// The following expression is in CNF:
+///  `(a OR b) AND (c OR d)`
+///
+/// The following is not in CNF:
+///  `(a AND b) OR c`.
+///
+/// But could be rewrite to a CNF expression:
+///  `(a OR c) AND (b OR c)`.
+///
+///
+/// # Example
+/// ```
+/// # use datafusion_expr::{col, lit};
+/// # use datafusion_optimizer::utils::cnf_rewrite;
+/// // （a=1 AND b=2）OR c = 3
+/// let expr1 = col("a").eq(lit(1)).and(col("b").eq(lit(2)));
+/// let expr2 = col("c").eq(lit(3));
+/// let expr = expr1.or(expr2);
+///
+///  //（a=1 or c=3）AND（b=2 or c=3）
+/// let expr1 = col("a").eq(lit(1)).or(col("c").eq(lit(3)));
+/// let expr2 = col("b").eq(lit(2)).or(col("c").eq(lit(3)));
+/// let expect = expr1.and(expr2);
+/// assert_eq!(expect, cnf_rewrite(expr));
+/// ```
+pub fn cnf_rewrite(expr: Expr) -> Expr {
+    // Find all exprs joined by OR
+    let disjuncts = split_binary(&expr, Operator::Or);
+
+    // For each expr, split now on AND
+    // A OR B OR C --> split each A, B and C
+    let disjunct_conjuncts: VecDeque<Vec<&Expr>> = disjuncts
+        .into_iter()
+        .map(|e| split_binary(e, Operator::And))
+        .collect::<VecDeque<_>>();
+
+    // Decide if we want to distribute the clauses. Heuristic is
+    // chosen to avoid creating huge predicates
+    let num_conjuncts = disjunct_conjuncts
+        .iter()
+        .fold(1usize, |sz, exprs| sz.saturating_mul(exprs.len()));
+
+    if disjunct_conjuncts.iter().any(|exprs| exprs.len() > 1)
+        && num_conjuncts < MAX_CNF_REWRITE_CONJUNCTS
+    {
+        let or_clauses = permutations(disjunct_conjuncts)
+            .into_iter()
+            // form the OR clauses( A OR B OR C ..)
+            .map(|exprs| disjunction(exprs.into_iter().cloned()).unwrap());
```

Comment:
seems we only clone the exprs once here🤔

Reply:
Yes that is my analysis too -- this code will clone exprs only if it rewrites the expression -- if not, it will not clone anything